### PR TITLE
SCP-2458: Extend InputField to include number inputs (using NumberFormat).

### PR DESCRIPTION
**Pull request by merivale**
I tried several things before settling on this solution, which I think is quite neat - some of the other things I tried were certainly a lot more complicated... :sweat_smile: Explanations of the various things are in the comments.

I've thrown in a dummy number input in the contract setup card, just for testing. Don't look too closely at the contract setup card itself, which continues to be visually broken - making that look like the new designs is the very next thing on my list.

Here's a screenshot showing the dummy number input (but you probably need to try it for yourself to get the full idea):

localhost_8009_ (3)

### CODE REVIEW

**Review by palas — COMMENTED** [review]
Looks good, but I don't understand the way it validates input.
Four thoughts:
- Would it be possible to filter out the "e" character?
- Why do characters disappear from the left side?
- For ADA at least I don't think it is super convenient to have to type all the zeroes, if you just want 40 ADA for example. But I am a bit biased towards my implementation ofc :)
- Also writing negative numbers seems a bit difficult

glitch

**Comment by merivale** [comment]
> Would it be possible to filter out the "e" character?

Hmm, that's funny - what browser were you using? In any case I think my second commit should have fixed this. Would you mind checking again?

> Why do characters disappear from the left side?

Not sure what you mean by this. Can you give an example?

> For ADA at least I don't think it is super convenient to have to type all the zeroes, if you just want 40 ADA for example. But I am a bit biased towards my implementation ofc :)

You don't have to type them - they're added for you automatically. Or have I misunderstood?

> Also writing negative numbers seems a bit difficult

Good point, I forgot about negative numbers! :flushed: I've fixed that with another commit.

**Comment by palas** [comment]
> > Would it be possible to filter out the "e" character?
> 
> Hmm, that's funny - what browser were you using? In any case I think my second commit should have fixed this. Would you mind checking again?
> 
> > Why do characters disappear from the left side?
> 
> Not sure what you mean by this. Can you give an example?
> 
> > For ADA at least I don't think it is super convenient to have to type all the zeroes, if you just want 40 ADA for example. But I am a bit biased towards my implementation ofc :)
> 
> You don't have to type them - they're added for you automatically. Or have I misunderstood?
> 
> > Also writing negative numbers seems a bit difficult
> 
> Good point, I forgot about negative numbers!  I've fixed that with another commit.

I used chromium, I think, I will give it another try, maybe record

File: marlowe-dashboard-client/src/Css.purs [hunk 1]
```diff
@@ -99,19 +97,31 @@ inputBaseNoFocus :: Array String
 inputBaseNoFocus = inputBase <> [ "focus:ring-0" ]
 
 input :: Boolean -> Array String
-input valid = inputBaseFocus <> [ "border", "bg-transparent" ] <> if valid then [ "border-black", "focus:border-black" ] else [ "border-red" ]
+input valid =
+  inputBaseFocus
+    <> [ "border-gray" ]
+    <> applyWhen (not valid) [ "border-red" ]
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/Css.purs:103]: I think an `if else` expression is better as there are no guarantees that the browser will pick the red border from the grey one if both are present (that has bitten me in the past)

File: marlowe-dashboard-client/src/Css.purs [hunk 2]
```diff
@@ -99,19 +97,31 @@ inputBaseNoFocus :: Array String
...
 
 inputNoFocus :: Boolean -> Array String
-inputNoFocus valid = inputBaseNoFocus <> if valid then [ "border-transparent" ] else [ "border-red" ]
+inputNoFocus valid =
+  inputBaseNoFocus
+    <> [ "border-gray" ]
+    <> applyWhen (not valid) [ "border-red" ]
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/Css.purs:109]: same

File: marlowe-dashboard-client/src/InputField/State.purs [hunk 3]
```diff
@@ -39,6 +54,33 @@ handleAction (SetValue value) =
...
+  let
+    decimals = case numberFormat of
+      DefaultFormat -> zero
+      DecimalFormat d _ -> d
+
+    bigIntegerValue = getBigIntegerValue decimals value
+
+    newValue = humanizeValue decimals bigIntegerValue
+
+    forcedDifferentValue = humanizeValue decimals $ bigIntegerValue + (BigInteger.fromInt 1)
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/InputField/State.purs:80]: I'm not sure I understand this part. I get the fact that in some cases we need to force halogen for a change, but isn't this modifying the value?

**Review by hrajchert — COMMENTED** [review]
The code looks good, but I think there might be something wrong with the functionality... I selected the dummy input, tried to enter 20 ADAs and as soon as I enter the 2 I lose focus and can't enter the zero.

input

File: marlowe-dashboard-client/src/Css.purs (hunk 1 above)
Comment (merivale) [reply marlowe-dashboard-client/src/Css.purs:103]: ah good point 👍

File: marlowe-dashboard-client/src/InputField/State.purs (hunk 3 above)
Comment (merivale) [reply marlowe-dashboard-client/src/InputField/State.purs:80]: I set the modified value and then immediately set it back to the right value. Do you know of a less hacky way to force the change?

**Comment by merivale** [comment]
Yeah, @palas noticed the same problem with the functionality. Bugger. I'll fiddle around some more tomorrow and see if there's a fix, otherwise we may have to accept it has to be a string input. 😖

**Comment by merivale** [comment]
Okay, after banging my head against the wall for a bit, I have come up with a new solution.

The first thing to note is that I'm no longer formatting the number `onValueInput` or `onKeyUp`, but `onBlur`. It would be nice if we could format it as you type, but I can't see a way of doing that without the problem of the cursor jumping to the end of the input when you don't want it to. (There is Pablo's nifty technique with `setSelectionRange`, but that only works if it's a string input, and then you lose all the nice native number input support. I think getting the native number support is worth this price.)

The second thing is that I've identified the problem that showed up previously with very large numbers. It was the fault of the NumberFormatter that we're using in the `Humanize` module (there's a comment about it in this latest commit).

And finally, as I was doing this I suddenly remembered about the CSS `focus-within` selector. Just the thing for a nice CSS-only implementation of the div-as-input that's needed to get the styling right. :tada:

**Comment by palas** [comment]
It seems quite robust now, but after trying really hard I managed to find another glitch:
When I write `0.00001` I get `0.100000'
I had the same problem, that is why I knew to try this.

Other than that there is just the CSS issue with the top line disappearing when you focus on the field

**Comment by merivale** [comment]
Phew, I think I've fixed it now. :sweat_smile:

**Comment by palas** [comment]
> Phew, I think I've fixed it now.

Bad news. I had almost given all hope but in the end I found another one 😎, if I write:
`-0.001`

The minus gets removed

**Comment by merivale** [comment]
Okay, I see what I did wrong there, and I've fixed that now. There is absolutely definitely no possible way you will ever find any more bugs in this implementation now. :grinning: Game on...

**Review by palas — APPROVED** [review]
Implementation looks good now :+1:
I was trying really hard to crash it and I started running on browser issues, so it is probably time to stop testing :)

**Pull request merged**